mount/mount_linux.go

```diff
@@ -38,6 +50,43 @@ func init() {
 	pagesize = os.Getpagesize()
 }
 
+// prepareIDMappedOverlay is a helper function to obtain
+// actual "lowerdir=..." mount options. It creates and
+// applies id mapping for each lowerdir.
+//
+// It returns:
+//  1. New options that include new "lowedir=..." mount option.
+//  2. "Clean up" function -- it should be called as a defer one before
+//     checking for error, because if do the second and avoid calling "clean up",
+//     you're going to have "dirty" setup -- there's no guarantee that those
+//     temporary mount points for lowedirs will be cleaned properly.
+//  3. Error -- nil if everything's fine, otherwise an error.
+func prepareIDMappedOverlay(usernsFd int, options []string) ([]string, func(), error) {
+	lowerIdx, lowerDirs := findOverlayLowerdirs(options)
+	if lowerIdx == -1 {
+		return options, nil, fmt.Errorf("failed to parse overlay lowerdir's from given options")
+	}
+
+	tmpLowerdirs, idMapCleanUp, err := doPrepareIDMappedOverlay(lowerDirs, usernsFd)
+	// schedule cleanup function first, so we're able to unmount/remove temporary
+	// directories if prepareIDMappedOverlay failed at the middle game.
+	cleanUp := func() {
+		idMapCleanUp()
+		if terr := os.RemoveAll(filepath.Clean(filepath.Join(tmpLowerdirs[0], ".."))); terr != nil {
```

Comment:
These are directories created by prepareIDMappedOverlaty. Shouldn't the be deleted by their cleanup function?

And maybe here we can delete the top level dir where they were created, if we created that and need to delete that too.

What do you think?

Reply:
>These are directories created by prepareIDMappedOverlaty. Shouldn't the be deleted by their cleanup function?

Yeah, should be better. I remember I have already tried this and there was some issue about it, but now I don't see anything that can stop doing this. Thanks for noticing! 

>And maybe here we can delete the top level dir where they were created, if we created that and need to delete that too.

That's actually what it does, it goes one level up and removes all nested and top level dirs